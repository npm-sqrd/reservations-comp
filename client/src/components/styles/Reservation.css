:local(.container) {
  margin-left: auto;
  margin-right: auto;
  padding-left: 20px;
  width: 720px;
  border-radius: 10px;
  background: #fff;
/*  box-shadow: 0 2px 8px #c1bfbf;*/
}

:local(.message) {
  padding: 4px;
  font-family: BrandonText,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  margin: 5px;
  padding-left: 15px;
  color: #333;
  padding-bottom:15px;
}
