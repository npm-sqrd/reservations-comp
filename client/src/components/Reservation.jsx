import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';

import SetName from './SetName';
import SearchParams from './SearchParams';
import TimeSlotSelector from './TimeSlotSelector';

import styles from './styles/Reservation.css';

import helper from '../lib/helper';

class Reservation extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      bookingsMadeToday: 0,
      availabilityInfo: [
        { time: 17, remaining: 0 },
        { time: 18, remaining: 0 },
        { time: 19, remaining: 0 },
        { time: 20, remaining: 0 },
        { time: 21, remaining: 0 },
      ],
      date: moment.tz('America/Los_Angeles').format('YYYY-MM-DD'),
      time: 19,
      name: '',
      party: 2,
      message: '',
    };

    this.setName = this.setName.bind(this);
    this.changeParty = this.changeParty.bind(this);
    this.requestReservation = this.requestReservation.bind(this);
    this.getAvailabilityInfo = this.getAvailabilityInfo.bind(this);
  }

  componentDidMount() {
    this.getAvailabilityInfo(this.state.date, this.state.time, this.state.party);
  }

  setName(name) {
    this.setState({ name });
  }

  getAvailabilityInfo(date, time, party) {
    helper.getReservationInfo(this.props.id, date, (err, data) => {
      if (!err) {
        this.setState({
          availabilityInfo: data.reservations,
          bookingsMadeToday: data.madeToday,
          party,
          date,
          time,
        });
      } else {
        this.setState({
          message: `Error getting availability info: ${err}`,
        });
      }
    });
  }

  changeParty(party) {
    this.setState({ party });
  }

  requestReservation(time) {
    if (this.state.name.length === 0) {
      this.setState({
        message: 'Enter your name before making a reservation!',
      });
    } else {
      helper.requestReservation(
        this.props.id, this.state.date, time, this.state.name, this.state.party,
        (err) => {
          if (!err) {
            this.setState({
              message: 'Your table has been saved!',
            });
            this.getAvailabilityInfo(this.state.date, this.state.time, this.state.party);
          } else {
            this.setState({
              message: 'Error making reservation',
            });
          }
        },
      );
    }
  }

  render() {
    const {
      name, bookingsMadeToday, availabilityInfo, message,
    } = this.state;

    return (
      <div className={styles.container}>
        <SetName
          name={name}
          clickHandler={this.setName}
        />
        <SearchParams
          clickHandler={this.getAvailabilityInfo}
          changeParty={this.changeParty}
        />
        <TimeSlotSelector
          party={this.state.party}
          bookingsMadeToday={bookingsMadeToday}
          availabilityInfo={availabilityInfo}
          clickHandler={this.requestReservation}
        />
        <div className={styles.message}>
          {message}
        </div>
      </div>
    );
  }
}

Reservation.propTypes = {
  id: PropTypes.number.isRequired,
};

export default Reservation;
