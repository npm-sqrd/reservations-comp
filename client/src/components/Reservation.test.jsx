import React from 'react';
import { shallow } from 'enzyme';

import Reservation from './Reservation';

jest.mock('../lib/helper');

describe('Reservation Component', () => {
  afterAll(() => {
    jest.unmock('../lib/helper');
  });

  test('should render correctly', () => {
    expect(shallow(<Reservation id={305} />)).toMatchSnapshot();
  });

  test('should setName correctly', () => {
    const component = shallow(<Reservation id={305} />);
    component.instance().setName('Sideshow Bob');
    expect(component.state().name).toBe('Sideshow Bob');
  });

  test('should changeParty correctly', () => {
    const component = shallow(<Reservation id={305} />);
    component.instance().changeParty(4);
    expect(component.state().party).toBe(4);
  });

  describe('getAvailaibilityInfo', () => {
    test('should update state correctly', () => {
      const component = shallow(<Reservation id={305} />);
      component.instance().getAvailabilityInfo('2018-04-01', 17, 2);

      const expected = [
        { time: 17, remaining: 10 },
        { time: 18, remaining: 5 },
        { time: 19, remaining: 0 },
        { time: 20, remaining: 2 },
        { time: 21, remaining: 20 },
      ]; // refer to ../lib/__mocks__/helper.js for this shape

      expect(component.state().bookingsMadeToday).toBe(42);
      expect(component.state().availabilityInfo).toMatchObject(expected);
    });

    test('should handle error correctly', () => {
      // requesting restaurantId !== 305 will result in error
      // in our mocked helper.getReservationInfo()
      const component = shallow(<Reservation id={100} />);
      component.instance().getAvailabilityInfo('2018-04-01', 17, 2);
      const errorMessage = 'Error getting availability info: Error!';
      expect(component.state().message).toBe(errorMessage);
    });
  });

  describe('requestReservation', () => {
    test('should require a name to be set', () => {
      // requesting restaurantId !== 305 will result in error
      // in our mocked helper.getReservationInfo()
      const component = shallow(<Reservation id={305} />);
      component.instance().requestReservation(19);
      const displayMessage = 'Enter your name before making a reservation!';
      expect(component.state().message).toBe(displayMessage);
    });

    test('should update state correctly', () => {
      // requesting restaurantId !== 305 will result in error
      // in our mocked helper.getReservationInfo()
      const component = shallow(<Reservation id={305} />);
      component.instance().setName('test name');
      component.instance().requestReservation(19);
      const displayMessage = 'Your table has been saved!';
      expect(component.state().message).toBe(displayMessage);
    });

    test('should handle error correctly', () => {
      // requesting restaurantId !== 305 will result in error
      // in our mocked helper.getReservationInfo()
      const component = shallow(<Reservation id={100} />);
      component.instance().setName('test name');
      component.instance().requestReservation(19);
      const errorMessage = 'Error making reservation';
      expect(component.state().message).toBe(errorMessage);
    });
  });
});
