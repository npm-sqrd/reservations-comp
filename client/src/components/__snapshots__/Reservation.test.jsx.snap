// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reservation Component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Reservation
    id={305}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SetName
          clickHandler={[Function]}
          name=""
        />,
        <SearchParams
          changeParty={[Function]}
          clickHandler={[Function]}
        />,
        <TimeSlotSelector
          availabilityInfo={
            Array [
              Object {
                "remaining": 10,
                "time": 17,
              },
              Object {
                "remaining": 5,
                "time": 18,
              },
              Object {
                "remaining": 0,
                "time": 19,
              },
              Object {
                "remaining": 2,
                "time": 20,
              },
              Object {
                "remaining": 20,
                "time": 21,
              },
            ]
          }
          bookingsMadeToday={42}
          clickHandler={[Function]}
          party={2}
        />,
        <div
          className={undefined}
        >
          
        </div>,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "clickHandler": [Function],
          "name": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "changeParty": [Function],
          "clickHandler": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "availabilityInfo": Array [
            Object {
              "remaining": 10,
              "time": 17,
            },
            Object {
              "remaining": 5,
              "time": 18,
            },
            Object {
              "remaining": 0,
              "time": 19,
            },
            Object {
              "remaining": 2,
              "time": 20,
            },
            Object {
              "remaining": 20,
              "time": 21,
            },
          ],
          "bookingsMadeToday": 42,
          "clickHandler": [Function],
          "party": 2,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": undefined,
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SetName
            clickHandler={[Function]}
            name=""
          />,
          <SearchParams
            changeParty={[Function]}
            clickHandler={[Function]}
          />,
          <TimeSlotSelector
            availabilityInfo={
              Array [
                Object {
                  "remaining": 10,
                  "time": 17,
                },
                Object {
                  "remaining": 5,
                  "time": 18,
                },
                Object {
                  "remaining": 0,
                  "time": 19,
                },
                Object {
                  "remaining": 2,
                  "time": 20,
                },
                Object {
                  "remaining": 20,
                  "time": 21,
                },
              ]
            }
            bookingsMadeToday={42}
            clickHandler={[Function]}
            party={2}
          />,
          <div
            className={undefined}
          >
            
          </div>,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "clickHandler": [Function],
            "name": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "changeParty": [Function],
            "clickHandler": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "availabilityInfo": Array [
              Object {
                "remaining": 10,
                "time": 17,
              },
              Object {
                "remaining": 5,
                "time": 18,
              },
              Object {
                "remaining": 0,
                "time": 19,
              },
              Object {
                "remaining": 2,
                "time": 20,
              },
              Object {
                "remaining": 20,
                "time": 21,
              },
            ],
            "bookingsMadeToday": 42,
            "clickHandler": [Function],
            "party": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": undefined,
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
